### PR TITLE
Added typing information for datetime64, timedelta64

### PR DESCRIPTION
**Pull request by bedder**
This is the final PR in the backlog of branches - this should hopefully be an easy one.

### CODE REVIEW

File: numpy-stubs/__init__.pyi [hunk 1]
```diff
@@ -402,14 +402,15 @@ class number(generic):
...
-class datetime64(_real_generic):
+class datetime64:
     @overload
     def __init__(
         self, _data: Union[datetime64, str, dt.datetime] = ..., _format: str = ...
     ) -> None: ...
     @overload
     def __init__(self, _data: int, _format: str) -> None: ...
-    def __sub__(self, other: datetime64) -> timedelta64: ...
+    def __add__(self, other: Union[timedelta64, int]) -> datetime64: ...
```
Comment (elvijs) [inline numpy-stubs/__init__.pyi]: `int`?!?!

I can see that it works, but why...?

File: numpy-stubs/__init__.pyi [hunk 2]
```diff
@@ -420,8 +421,21 @@ class int64(signedinteger): ...
...
 class timedelta64(signedinteger):
     def __init__(self, _data: Any = ..., _format: str = ...) -> None: ...
-    def __add__(self, other: timedelta64) -> timedelta64: ...
-    def __sub__(self, other: timedelta64) -> timedelta64: ...
+    @overload
+    def __add__(self, other: Union[timedelta64, int]) -> timedelta64: ...
+    @overload
+    def __add__(self, other: datetime64) -> datetime64: ...
+    def __sub__(self, other: Union[timedelta64, int]) -> timedelta64: ...
+    if sys.version_info[0] < 3:
```
Comment (elvijs) [inline numpy-stubs/__init__.pyi:432]:
```suggestion
    if sys.version_info.major < 3:
```

File: numpy-stubs/__init__.pyi [hunk 3]
```diff
@@ -402,14 +402,15 @@ class number(generic):
...
+class datetime64:
     @overload
     def __init__(
         self, _data: Union[datetime64, str, dt.datetime] = ..., _format: str = ...
     ) -> None: ...
     @overload
     def __init__(self, _data: int, _format: str) -> None: ...
-    def __sub__(self, other: datetime64) -> timedelta64: ...
+    def __add__(self, other: Union[timedelta64, int]) -> datetime64: ...
+    def __sub__(self, other: Union[timedelta64, datetime64, int]) -> timedelta64: ...
```
Comment (elvijs) [inline numpy-stubs/__init__.pyi]: Do we need an overload here? `datetime64 - datetime64 = timedelta64`, but `datetime64 - timedelta/int = datetime64`, right?

**Review by elvijs — COMMENTED** [review]
Overall, looks good! The only potential issue is `datetime64.__sub__`: I think it might make sense to add some overloads there. See the comments below and let me know if you disagree

File: numpy-stubs/__init__.pyi (hunk 1 above)
Comment (bedder) [reply numpy-stubs/__init__.pyi]: You're going to love this! It adds/subtracts `datetime_variable.dtype.kind`, leading to some fun things like:
```
In[22]:  np.datetime64('2001') == np.datetime64('2001-01-01')
Out[22]: True
In[23]:  (np.datetime64('2001') + 1) == (np.datetime64('2001-01-01') + 1)
Out[23]: False
```

File: numpy-stubs/__init__.pyi (hunk 3 above)
Comment (bedder) [reply numpy-stubs/__init__.pyi]: You're totally correct - I will add the overloads.

File: numpy-stubs/__init__.pyi (hunk 1 above)
Comment (elvijs) [reply numpy-stubs/__init__.pyi]: :man_facepalming:

**Review by elvijs — APPROVED** [review]
The changes look good!

File: numpy-stubs/__init__.pyi (hunk 2 above)
Comment (bedder) [reply numpy-stubs/__init__.pyi:432]: `flake8` doesn't like those - it seems like mypy only supports (or at least flake8 thinks that mypy only supports) changing types when switching on the system version in the old style...

Comment (elvijs) [reply numpy-stubs/__init__.pyi:432]: ;(

**Pull request merged**